cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "indexing/construction/detail/utilities.cuh"
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/spatial_join.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/gather.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/reduce.h>
+
+#include <vector>
+
+namespace cuspatial {
+namespace detail {
+namespace {
+
+static uint8_t const max_warps_per_block  = 32;
+static uint8_t const num_threads_per_warp = 32;
+static uint32_t const threads_per_block   = 256;
+
+template <typename T>
+__global__ void quad_pip_phase1_kernel(uint32_t const *pq_poly_id,
+                                       uint32_t const *pq_quad_id,
+                                       uint32_t const *block_offset,
+                                       uint32_t const *block_length,
+                                       cudf::column_device_view const quad_offsets,
+                                       cudf::column_device_view const point_indices,
+                                       cudf::column_device_view const point_x,
+                                       cudf::column_device_view const point_y,
+                                       cudf::column_device_view const poly_offsets,
+                                       cudf::column_device_view const ring_offsets,
+                                       cudf::column_device_view const poly_points_x,
+                                       cudf::column_device_view const poly_points_y,
+                                       uint32_t *num_hits)
+{
+  __shared__ uint32_t quad_idx, poly_idx, num_point, first_pos, qpos, num_adjusted;
+
+  // assume # of points/threads no more than num_threads_per_warp * max_warps_per_block (32*32)
+  __shared__ uint16_t data[max_warps_per_block];
```

Comment:
`data` is not a very helpful name. I think this is `warp_num_hits` or something?